vividus-plugin-web-app/src/main/java/org/vividus/ui/web/listener/WebSourceCodePublishingOnFailureListener.java

```diff
@@ -28,34 +29,71 @@
 import org.vividus.selenium.IWebDriverProvider;
 import org.vividus.ui.context.IUiContext;
 import org.vividus.ui.listener.AbstractSourceCodePublishingOnFailureListener;
+import org.vividus.ui.web.action.CssSelectorFactory;
+import org.vividus.ui.web.action.WebJavascriptActions;
 
 public class WebSourceCodePublishingOnFailureListener extends AbstractSourceCodePublishingOnFailureListener
 {
     private static final Logger LOGGER = LoggerFactory.getLogger(WebSourceCodePublishingOnFailureListener.class);
 
     private final IUiContext uiContext;
+    private final WebJavascriptActions webJavascriptActions;
 
     protected WebSourceCodePublishingOnFailureListener(IAttachmentPublisher attachmentPublisher, IWebDriverProvider
-             webDriverProvider, IUiContext uiContext)
+             webDriverProvider, IUiContext uiContext, WebJavascriptActions webJavascriptActions)
     {
         super(attachmentPublisher, webDriverProvider, "HTML");
         this.uiContext = uiContext;
+        this.webJavascriptActions = webJavascriptActions;
     }
 
     @Override
-    protected Optional<String> getSourceCode()
+    protected Map<String, String> getSourceCode()
     {
         SearchContext searchContext = uiContext.getSearchContext();
         String sourceCode = null;
-        if (searchContext instanceof WebElement)
+        boolean elementInContext = searchContext instanceof WebElement;
+        if (elementInContext)
         {
             sourceCode = getElementSource(searchContext);
         }
         else if (searchContext instanceof WebDriver)
         {
             sourceCode = ((WebDriver) searchContext).getPageSource();
         }
-        return Optional.ofNullable(sourceCode);
+        Map<String, String> sources = new LinkedHashMap<>();
+        if (sourceCode != null)
+        {
+            sources.put(APPLICATION_SOURCE_CODE, sourceCode);
+        }
+        sources.putAll(getShadowDomSourceCode(elementInContext, searchContext));
```

Comment:
how is performance affected?

Reply:
```
Lifecycle:
Examples:
{transformer=ITERATING, limit=200}

Scenario: Verify method DEBUG is supported; accidental space is trimmed in httpMethod enum
Given I am on a page with the URL 'http://localhost:8080/shadowDom.html'
When I change context to element located `xpath(//body)`
Then number of elements found by `id(123)` is = `1`
```
in 20 threads

6m 48s - shadow dom
6m 33s - master